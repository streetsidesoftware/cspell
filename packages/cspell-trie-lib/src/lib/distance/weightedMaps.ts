import { SuggestionCostMapDef } from '../models/suggestionCostsDef';

export type WeightedRepMapTrie = Record<string, WeightedRepTrieNode>;

interface WeightedRepTrieNode {
    /** The nested Trie nodes */
    r?: WeightedRepMapTrie | undefined;
    /** The cost to replace */
    rep?: number | undefined;
    /** The cost to swap */
    swap?: number | undefined;
}

/**
 * Costs are minimized while penalties are maximized.
 */
interface Cost {
    /**
     * The cost of an operation
     * `c'' = min(c, c')`
     */
    c?: number | undefined;
    /**
     * The penalties applied
     * `p'' = max(p, p')`
     */
    p?: number | undefined;
}

interface TrieCost extends Cost {
    /** nested trie nodes */
    n?: Record<string, TrieCost>;
}

interface TrieTrieCost {
    /** nested trie nodes */
    n?: Record<string, TrieTrieCost>;
    /** root of cost trie */
    t?: Record<string, TrieCost>;
}

export interface CostPosition {
    // Starting from word
    a: string;
    // offset within `a`
    ai: number;
    // Goal word
    b: string;
    // offset within `b`
    bi: number;
    // accumulated cost to this point.
    c: number;
    // accumulated penalties to this point.
    p: number;
}

export interface WeightMap {
    readonly insDel: TrieCost;
    readonly replace: TrieTrieCost;
    readonly swap: TrieTrieCost;

    calcInsDelCosts(pos: CostPosition): Iterable<CostPosition>;
    calcSwapCosts(pos: CostPosition): Iterable<CostPosition>;
    calcReplaceCosts(pos: CostPosition): Iterable<CostPosition>;
}

export function createWeightMap(...defs: SuggestionCostMapDef[]): WeightMap {
    const map = _createWeightMap();
    _addDefToWeightMap(map, ...defs);
    return map;
}

export function addDefToWeightMap(
    map: WeightMap,
    def: SuggestionCostMapDef,
    ...defs: SuggestionCostMapDef[]
): WeightMap;
export function addDefToWeightMap(map: WeightMap, ...defs: SuggestionCostMapDef[]): WeightMap {
    return _addDefToWeightMap(map, ...defs);
}

function _addDefToWeightMap(map: WeightMap, ...defs: SuggestionCostMapDef[]): WeightMap {
    function addSet(set: string[], def: SuggestionCostMapDef) {
        addSetToTrieCost(map.insDel, set, def.insDel, def.penalty);
        addSetToTrieTrieCost(map.replace, set, def.replace, def.penalty);
        addSetToTrieTrieCost(map.swap, set, def.swap, def.penalty);
    }

    for (const def of defs) {
        const mapSets = splitMap(def);
        mapSets.forEach((s) => addSet(s, def));
    }
    return map;
}
function _createWeightMap(): WeightMap {
    return new _WeightedMap();
}

function lowest(a: number | undefined, b: number | undefined): number | undefined {
    if (a === undefined) return b;
    if (b === undefined) return a;
    return a <= b ? a : b;
}

function highest(a: number | undefined, b: number | undefined): number | undefined {
    if (a === undefined) return b;
    if (b === undefined) return a;
    return a >= b ? a : b;
}

export function* splitMapSubstringsIterable(map: string): Iterable<string> {
    let seq = '';
    let mode = 0;
    for (const char of map) {
        if (mode && char === ')') {
            yield seq;
            mode = 0;
            continue;
        }
        if (mode) {
            seq += char;
            continue;
        }
        if (char === '(') {
            mode = 1;
            seq = '';
            continue;
        }
        yield char;
    }
}

export function splitMapSubstrings(map: string): string[] {
    return [...splitMapSubstringsIterable(map)];
}

/**
 * Splits a WeightedMapDef.map
 * @param map
 */
function splitMap(def: Pick<SuggestionCostMapDef, 'map'>): string[][] {
    const { map } = def;

    const sets = map.split('|');
    return sets.map(splitMapSubstrings).filter((s) => s.length > 0);
}

function addToTrieCost(trie: TrieCost, str: string, cost: number, penalties: number | undefined): void {
    if (!str) return;
    let t = trie;
    for (const c of str) {
        const n = (t.n = t.n || Object.create(null));
        t = n[c] = n[c] || Object.create(null);
    }
    t.c = lowest(t.c, cost);
    t.p = highest(t.p, penalties);
}

function addToTrieTrieCost(
    trie: TrieTrieCost,
    left: string,
    right: string,
    cost: number,
    penalties: number | undefined
): void {
    let t = trie;
    for (const c of left) {
        const n = (t.n = t.n || Object.create(null));
        t = n[c] = n[c] || Object.create(null);
    }
    const trieCost = (t.t = t.t || Object.create(null));
    addToTrieCost(trieCost, right, cost, penalties);
}

function addSetToTrieCost(trie: TrieCost, set: string[], cost: number | undefined, penalties: number | undefined) {
    if (cost === undefined) return;
    for (const str of set) {
        addToTrieCost(trie, str, cost, penalties);
    }
}

function addSetToTrieTrieCost(
    trie: TrieTrieCost,
    set: string[],
    cost: number | undefined,
    penalties: number | undefined
) {
    if (cost === undefined) return;
    for (const left of set) {
        for (const right of set) {
            if (left === right) continue;
            addToTrieTrieCost(trie, left, right, cost, penalties);
        }
    }
}

function* searchTrieNodes<T extends { n?: Record<string, T> }>(trie: T, str: string, i: number) {
    const len = str.length;

    for (let n = trie.n; i < len && n; ) {
        const t = n[str[i]];
        if (!t) return;
        ++i;
        yield { i, t };
        n = t.n;
    }
}

function* walkTrieNodes<T extends { n?: Record<string, T> }>(
    t: T | undefined,
    s: string
): Generator<{ s: string; t: T }> {
    if (!t) return;

    yield { s, t };

    if (!t.n) return;

    for (const [k, v] of Object.entries(t.n)) {
        yield* walkTrieNodes(v, s + k);
    }
}

function* walkTrieCost(trie: TrieCost): Generator<{ s: string; c: number; p: number | undefined }> {
    for (const { s, t } of walkTrieNodes(trie, '')) {
        if (t.c) {
            yield { s, c: t.c, p: t.p };
        }
    }
}

function* walkTrieTrieCost(trie: TrieTrieCost): Generator<{ a: string; b: string; c: number; p: number | undefined }> {
    for (const { s: a, t } of walkTrieNodes(trie, '')) {
        if (t.t) {
            for (const { s: b, c, p } of walkTrieCost(t.t)) {
                yield { a, b, c, p };
            }
        }
    }
}

interface MatchTrieCost {
    i: number;
    c: number;
    p: number;
}

function* findTrieCostPrefixes(trie: TrieCost, str: string, i: number): Iterable<MatchTrieCost> {
    for (const n of searchTrieNodes(trie, str, i)) {
        const { c, p } = n.t;
        if (c !== undefined) {
            yield { i: n.i, c, p: p || 0 };
        }
    }
}

interface MatchTrieTrieCost {
    i: number;
    t: TrieCost;
}

function* findTrieTrieCostPrefixes(trie: TrieTrieCost, str: string, i: number): Iterable<MatchTrieTrieCost> {
    for (const n of searchTrieNodes(trie, str, i)) {
        const t = n.t.t;
        if (t !== undefined) {
            yield { i: n.i, t };
        }
    }
}

class _WeightedMap implements WeightMap {
    insDel: TrieCost = {};
    replace: TrieTrieCost = {};
    swap: TrieTrieCost = {};

    *calcInsDelCosts(pos: CostPosition): Iterable<CostPosition> {
        const { a, ai, b, bi, c, p } = pos;
        for (const del of findTrieCostPrefixes(this.insDel, a, ai)) {
            yield { a, b, ai: del.i, bi, c: c + del.c, p: p + del.p };
        }
        for (const ins of findTrieCostPrefixes(this.insDel, b, bi)) {
            yield { a, b, ai, bi: ins.i, c: c + ins.c, p: p + ins.p };
        }
    }

    *calcReplaceCosts(pos: CostPosition): Iterable<CostPosition> {
        // Search for matching substrings in `a` to be replaced by
        // matching substrings from `b`. All substrings start at their
        // respective `ai`/`bi` positions.
        const { a, ai, b, bi, c, p } = pos;
        for (const del of findTrieTrieCostPrefixes(this.replace, a, ai)) {
            for (const ins of findTrieCostPrefixes(del.t, b, bi)) {
                yield { a, b, ai: del.i, bi: ins.i, c: c + ins.c, p: p + ins.p };
            }
        }
    }

    *calcSwapCosts(pos: CostPosition): Iterable<CostPosition> {
        const { a, ai, b, bi, c, p } = pos;
        const swap = this.swap;

        for (const left of findTrieTrieCostPrefixes(swap, a, ai)) {
            for (const right of findTrieCostPrefixes(left.t, a, left.i)) {
                const sw = a.slice(left.i, right.i) + a.slice(ai, left.i);
                if (b.slice(bi).startsWith(sw)) {
                    const len = sw.length;
                    yield { a, b, ai: ai + len, bi: bi + len, c: c + right.c, p: p + right.p };
                }
            }
        }
    }
}

function prettyPrintInsDel(trie: TrieCost, pfx = '', indent = '  '): string {
    function* walk() {
        for (const { s, c, p } of walkTrieCost(trie)) {
            const pm = p ? ` + ${p}` : '';
            yield indent + `(${s}) = ${c}${pm}`;
        }
    }
    return ['InsDel:', ...[...walk()].sort()].map((line) => pfx + line + '\n').join('');
}

export function prettyPrintReplace(trie: TrieTrieCost, pfx = '', indent = '  '): string {
    function* walk() {
        for (const { a, b, c, p } of walkTrieTrieCost(trie)) {
            const pm = p ? ` + ${p}` : '';
            yield indent + `(${a}) -> (${b}) = ${c}${pm}`;
        }
    }
    return ['Replace:', ...[...walk()].sort()].map((line) => pfx + line + '\n').join('');
}

export function prettyPrintSwap(trie: TrieTrieCost, pfx = '', indent = '  '): string {
    function* walk() {
        for (const { a, b, c, p } of walkTrieTrieCost(trie)) {
            const pm = p ? ` + ${p}` : '';
            yield indent + `(${a}) <-> (${b}) = ${c}${pm}`;
        }
    }
    return ['Swap:', ...[...walk()].sort()].map((line) => pfx + line + '\n').join('');
}

export function prettyPrintWeightMap(map: WeightMap): string {
    return [prettyPrintInsDel(map.insDel), prettyPrintReplace(map.replace), prettyPrintSwap(map.swap)].join('\n');
}

export function lookupReplaceCost(map: WeightMap, a: string, b: string): undefined | number {
    const trie = map.replace;

    let tt: TrieTrieCost | undefined = trie;
    for (let ai = 0; ai < a.length && tt; ++ai) {
        tt = tt.n?.[a[ai]];
    }
    if (!tt) return undefined;

    let t: TrieCost | undefined = tt.t;
    for (let bi = 0; bi < b.length && t; ++bi) {
        t = t.n?.[b[bi]];
    }

    return t?.c;
}

export const __testing__ = {
    splitMap,
    splitMapSubstrings,
    findTrieCostPrefixes,
    findTrieTrieCostPrefixes,
};
