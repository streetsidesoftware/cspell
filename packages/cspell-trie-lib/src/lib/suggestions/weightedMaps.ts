export type WeightedMapTrie = Record<string, WeightedMapTrieNode>;

interface WeightedMapTrieNode {
    /** The nested Tri nodes */
    t?: WeightedMapTrie | undefined;
    /** the cost to insert/delete this string */
    insDel?: number | undefined;
    /** the related replacement Tri Map */
    r?: WeightedRepMapTrie | undefined;
}

export type WeightedRepMapTrie = Record<string, WeightedRepTrieNode>;

interface WeightedRepTrieNode {
    /** The nested Trie nodes */
    r?: WeightedRepMapTrie | undefined;
    /** The cost to replace */
    rep?: number | undefined;
    /** The cost to swap */
    swap?: number | undefined;
}

export interface WeightedMapDef {
    /**
     * The set of substrings to map, these are generally single character strings.
     *
     * Multiple sets can be defined by using a `|` to separate them.
     *
     * Example: `"eéê|aåá"` contains two different sets.
     *
     * To add a multi-character substring use `()`.
     *
     * Example: `"f(ph)(gh)"` results in the following set: `f`, `ph`, `gh`.
     */
    map: string;
    /** The cost to insert/delete one of the substrings in the map. Note: insert/delete costs are symmetrical. */
    insDel?: number;
    /**
     * The cost to replace of of the substrings in the map with another substring in the map.
     * Example: Map['a', 'i']
     * This would be the cost to substitute `a` with `i`: Like `bat` to `bit` or the reverse.
     */
    replace?: number;
    /**
     * The cost to swap two adjacent substrings found in the map.
     * Example: Map['e', 'i']
     * This represents the cost to change `ei` to `ie` or the reverse.
     */
    swap?: number;
}

export function buildWeightedMapTrie(defs: WeightedMapDef[]): WeightedMapTrie {
    const trie: WeightedMapTrie = createMapTrie();
    defs.forEach((def) => addWeightedDefMapToTrie(def, trie));
    return trie;
}

/**
 * Add weighted map definitions to a WeightedMapTrie
 * @param def - the def to add
 * @param trie - the trie to add it to. NOTE: this trie is modified!
 * @returns the modified trie
 */
export function addWeightedDefMapToTrie(def: WeightedMapDef, trie: WeightedMapTrie = createMapTrie()): WeightedMapTrie {
    const mapSets = splitMap(def);

    function addRepToNode(mapSet: string[], n: WeightedMapTrieNode) {
        const root = createRepTrieNode((n.r = n.r || createRepTrie()));
        for (const s of mapSet) {
            let n = root;
            for (const c of s) {
                const r = (n.r = n.r || createRepTrie());
                n = r[c] = r[c] || createRepTrieNode();
            }
            addWeightsToRepNode(n, def);
        }
    }

    function addSet(mapSet: string[]) {
        const r = createMapTrieNode(trie);
        for (const s of mapSet) {
            let n = r;
            for (const c of s) {
                const t = (n.t = n.t || createMapTrie());
                n = t[c] = t[c] || createMapTrieNode();
            }
            addWeightsToNode(n, def);
            if (def.replace !== undefined || def.swap !== undefined) {
                addRepToNode(mapSet, n);
            }
        }
    }

    mapSets.forEach(addSet);

    return trie;
}

function createMapTrie(): WeightedMapTrie {
    return Object.create(null);
}

function createMapTrieNode(t?: WeightedMapTrie): WeightedMapTrieNode {
    const n: WeightedMapTrieNode = {};
    return assignIfDefined(n, 't', t);
}

function createRepTrie(): WeightedRepMapTrie {
    return Object.create(null);
}

function createRepTrieNode(r?: WeightedRepMapTrie): WeightedRepTrieNode {
    const n: WeightedRepTrieNode = {};
    return assignIfDefined(n, 'r', r);
}

function addWeightsToNode(n: WeightedMapTrieNode, def: WeightedMapDef) {
    assignIfDefined(n, 'insDel', lowest(n.insDel, def.insDel));
}

function addWeightsToRepNode(n: WeightedRepTrieNode, def: WeightedMapDef) {
    assignIfDefined(n, 'rep', lowest(n.rep, def.replace));
    assignIfDefined(n, 'swap', lowest(n.swap, def.swap));
}

function lowest(a: number | undefined, b: number | undefined): number | undefined {
    if (a === undefined) return b;
    if (b === undefined) return a;
    return a <= b ? a : b;
}

/**
 * Splits a WeightedMapDef.map
 * @param map
 */
function splitMap(def: WeightedMapDef): string[][] {
    const { map } = def;

    const sets = map.split('|');
    return sets.map(splitMapSubstrings);
}

function splitMapSubstrings(map: string): string[] {
    const values = [];
    const len = map.length;

    for (let i = 0; i < len; ++i) {
        const c = map[i];
        if (c !== '(') {
            values.push(c);
            continue;
        }
        const s = i + 1;
        while (map[++i] !== ')' && i < len) {
            // empty
        }
        values.push(map.slice(s, i));
    }

    return values;
}

function assignIfDefined<T, K extends keyof T>(r: T, key: K, v: T[K] | undefined): T {
    if (v === undefined) return r;
    r[key] = v;
    return r;
}

export const __testing__ = {
    splitMap,
    splitMapSubstrings,
};
