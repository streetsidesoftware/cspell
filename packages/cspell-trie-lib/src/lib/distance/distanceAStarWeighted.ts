import { PairingHeap } from '../utils/PairingHeap';
import { WeightMap } from './weightedMaps';

/**
 * Calculate the edit distance between two words using an A* algorithm.
 *
 * Using basic weights, this algorithm has the same results as the Damerau-Levenshtein algorithm.
 */
export function distanceAStarWeighted(a: string, b: string, map: WeightMap, cost = 100): number {
    const aN = a.length;
    const bN = b.length;

    const candidates = new PairingHeap(compare);

    candidates.add({ ai: 0, bi: 0, c: 0 });

    function opSub(n: Node) {
        const { ai, bi, c } = n;
        if (ai < aN && bi < bN) {
            const cc = a[ai] === b[bi] ? c : c + cost;
            candidates.add({ ai: ai + 1, bi: bi + 1, c: cc });
        }
    }

    function opIns(n: Node) {
        const { ai, bi, c } = n;
        if (bi < bN) {
            candidates.add({ ai: ai, bi: bi + 1, c: c + cost });
        }
    }

    function opDel(n: Node) {
        const { ai, bi, c } = n;
        if (ai < aN) {
            candidates.add({ ai: ai + 1, bi: bi, c: c + cost });
        }
    }

    function opSwap(n: Node) {
        const { ai, bi, c } = n;
        if (a[ai] === b[bi + 1] && a[ai + 1] === b[bi]) {
            candidates.add({ ai: ai + 2, bi: bi + 2, c: c + cost });
        }
    }

    function opMap(n: Node) {
        const { ai, bi, c } = n;
        const pos = { a, b, ai, bi, c };
        const costCalculations = [map.calcInsDelCosts(pos), map.calcSwapCosts(pos), map.calcReplaceCosts(pos)];
        costCalculations.forEach((iter) => {
            for (const nn of iter) {
                candidates.add(nn);
            }
        });
    }

    let best: Node | undefined;
    // const bc2 = 2 * bc;
    while ((best = candidates.dequeue())) {
        if (best.ai === aN && best.bi === bN) break;

        opSwap(best);
        opIns(best);
        opDel(best);
        opMap(best);
        opSub(best);
    }

    // istanbul ignore else
    return best ? best.c : -1;
}

interface Pos {
    /** the offset in string `a` */
    ai: number;
    /** the offset in string `b` */
    bi: number;
}

interface Node extends Pos {
    /** the current cost */
    c: number;
}

function compare(a: Node, b: Node): number {
    // Choose lowest cost or farthest Manhattan distance.
    return a.c - b.c || b.ai + b.bi - a.ai - a.bi;
}
