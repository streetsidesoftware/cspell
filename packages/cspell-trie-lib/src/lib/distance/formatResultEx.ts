import type { ExResult } from './distanceAStarWeighted';
import { distanceAStarWeightedEx } from './distanceAStarWeighted';
import type { WeightMap } from './weightedMaps';

function pL(s: string, w: number) {
    const strWidth = vizWidth(s);
    const w0 = Math.max(0, w - strWidth);
    return ' '.repeat(w0) + s;
}

function pR(s: string, w: number) {
    const strWidth = vizWidth(s);
    const w0 = Math.max(0, w - strWidth);
    return s + ' '.repeat(w0);
}

function vizWidth(s: string) {
    const r = s.replace(/[\u0300-\u036F\u007f-\u009f]/gu, '');
    let i = 0;
    for (const c of r) {
        i += c.length;
    }
    return i;
}

export function formatExResult(ex: ExResult | undefined): string {
    if (!ex) return '<undefined>';

    const { cost, segments, penalty } = ex;
    const asString = segments.map(({ a, b, c, p }) => ({
        a: `<${a}>`,
        b: `<${b}>`,
        c: c.toString(10),
        p: p.toString(10),
    }));
    asString.push({
        a: '',
        b: '',
        c: ' = ' + segments.reduce((sum, { c }) => sum + c, 0).toString(10),
        p: ' = ' + segments.reduce((sum, { p }) => sum + p, 0).toString(10),
    });
    const parts = asString.map(({ a, b, c, p }) => ({
        a,
        b,
        c,
        p,
        w: Math.max(vizWidth(a), vizWidth(b), vizWidth(c), vizWidth(p)),
    }));
    const a = 'a: |' + parts.map(({ a, w }) => pR(a, w)).join('|') + '|';
    const b = 'b: |' + parts.map(({ b, w }) => pR(b, w)).join('|') + '|';
    const c = 'c: |' + parts.map(({ c, w }) => pL(c, w)).join('|') + '|';
    const p = 'p: |' + parts.map(({ p, w }) => pL(p, w)).join('|') + '|';
    const penaltyMsg = penalty ? `[+${penalty}]` : '';
    return `<${ex.a.slice(1, -1)}> -> <${ex.b.slice(1, -1)}> (${cost - penalty}${penaltyMsg})\n${[a, b, c, p].join(
        '\n'
    )}\n`;
}

export function formattedDistance(wordA: string, wordB: string, weightMap: WeightMap, cost?: number) {
    const x = distanceAStarWeightedEx(wordA, wordB, weightMap, cost);
    return formatExResult(x);
}
