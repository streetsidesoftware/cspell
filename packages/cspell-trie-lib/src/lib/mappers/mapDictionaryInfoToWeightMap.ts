import { SuggestionCostMapDef } from '..';
import { addAdjustment, createWeightMap, PenaltyAdjustment, WeightMap } from '../distance/weightedMaps';
import type { DictionaryInformation } from '../models/DictionaryInformation';
import { mapDictionaryInformation, mapDictionaryInformationToAdjustment } from './mapDictionaryInfo';

const defaultDefs: SuggestionCostMapDef[] = [
    {
        map: '1234567890-.',
        insDel: 1,
        penalty: 200,
    },
];

const defaultAdjustments: PenaltyAdjustment[] = [
    {
        id: 'compound-case-change',
        regexp: /\p{Ll}∙\p{Lu}/gu,
        penalty: 1000,
    },
    {
        id: 'short-compounds-1',
        regexp: /∙.{1,2}(?=∙|$)/gu,
        penalty: 100,
    },
    {
        id: 'short-compounds-3',
        regexp: /∙.{3}(?=∙|$)/gu,
        penalty: 50,
    },
];

export function mapDictionaryInformationToWeightMap(dictInfo: DictionaryInformation): WeightMap {
    const defs = mapDictionaryInformation(dictInfo).concat(defaultDefs);
    const adjustments = mapDictionaryInformationToAdjustment(dictInfo);
    const map = createWeightMap(...defs);
    addAdjustment(map, ...defaultAdjustments, ...adjustments);
    return map;
}

export const __testing__ = {};
