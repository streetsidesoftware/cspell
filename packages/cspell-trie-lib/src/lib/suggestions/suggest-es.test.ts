import type { DictionaryInformation } from '@cspell/cspell-types';
import { describe, expect, test } from 'vitest';

import { readTrieFromConfig } from '../../test/dictionaries.test.helper.js';
import { distanceAStarWeightedEx } from '../distance/distanceAStarWeighted.js';
import { formatExResult } from '../distance/formatResultEx.js';
import type { WeightMap } from '../distance/index.js';
import { createWeightCostCalculator } from '../distance/weightedMaps.js';
import { mapDictionaryInformationToWeightMap } from '../mappers/mapDictionaryInfoToWeightMap.js';
import { parseLinesToDictionaryLegacy } from '../SimpleDictionaryParser.js';

function getTrie() {
    return readTrieFromConfig('@cspell/dict-es-es/cspell-ext.json');
}

describe('Validate Spanish Suggestions', () => {
    // cspell:locale en,es
    // cspell:ignore Carmjen
    // cspell:disableCaseSensitive
    // cspell:ignore barcelona carmjen nino
    test.each`
        word           | ignoreCase | expectedWords
        ${'carmjen'}   | ${false}   | ${['carmen', 'carmene', 'carmena', 'carmené', 'carmeno', 'carmenó', 'carmenen']}
        ${'carmjen'}   | ${true}    | ${['carmen', 'carmene', 'carmena', 'carmené', 'carmeno', 'carmenó', 'carmenen']}
        ${'carmen'}    | ${true}    | ${['carmen', 'carmene', 'carmena', 'carmené', 'carmeno', 'carmenó']}
        ${'carmen'}    | ${false}   | ${['carmen', 'carmene', 'carmena', 'carmené', 'carmeno', 'carmenó']}
        ${'cafe'}      | ${false}   | ${['café', 'cafés', 'cafre', 'cabe', 'cace', 'cale', 'cape', 'case', 'cate', 'cave']}
        ${'niño'}      | ${false}   | ${['niño', 'niños', 'niña', 'niñeo']}
        ${'nino'}      | ${false}   | ${['niño', 'ninfo', 'niños', 'nido', 'niña', 'nito', 'niñeo']}
        ${'barcelona'} | ${false}   | ${['Barcelona', 'parcelan', 'abretona', 'barceos', 'barcarola']}
        ${'Mexico'}    | ${false}   | ${['México', 'mexica', 'medico', 'medicó', 'médico']}
    `(
        'Tests suggestions "$word" ignoreCase: $ignoreCase',
        async ({ word, ignoreCase, expectedWords }) => {
            const trie = await getTrie();
            const suggestions = trie.suggest(word, { numSuggestions: 4, ignoreCase });
            expect(suggestions).toEqual(expectedWords);
        },
        { timeout: 5000 },
    );

    test.each`
        word      | ignoreCase | expectedWords
        ${'niño'} | ${false}   | ${[c('niño', 0), c('niños', 95), c('niña', 96), c('niñeo', 96)]}
        ${'nino'} | ${false}   | ${[c('niño', 1), c('ninfo', 96), c('niños', 96), c('nido', 97), c('niña', 97), c('nito', 97), c('niñeo', 97)]}
    `(
        'Tests suggestions "$word" ignoreCase: $ignoreCase',
        async ({ word, ignoreCase, expectedWords }) => {
            const trie = await getTrie();
            const results = trie.suggestWithCost(word, { numSuggestions: 4, ignoreCase });
            expect(results).toEqual(expectedWords);
        },
        { timeout: 5000 },
    );

    test.each`
        word      | ignoreCase | expectedWords
        ${'niño'} | ${false}   | ${[c('niño', 0), c('niños', 95), c('niña', 96), c('niñeo', 96), c('nido', 97), c('dino', 100)]}
        ${'nino'} | ${false}   | ${[c('niño', 1), c('niños', 96), c('nido', 97), c('niña', 97), c('niñeo', 97), c('dino', 99)]}
    `('Tests suggestions "$word" ignoreCase: $ignoreCase', ({ word, ignoreCase, expectedWords }) => {
        const trie = trieSimple();
        const results = trie.suggestWithCost(word, { numSuggestions: 10, ignoreCase });
        expect(results).toEqual(expectedWords);
    });

    // cspell:ignore nïño
    test.each`
        word      | ignoreCase | expectedWords
        ${'niño'} | ${false}   | ${[c('niño', 0), c('niños', 50), c('niña', 75), c('niñeo', 75)]}
        ${'nïño'} | ${false}   | ${[c('niño', 1), c('niños', 51), c('niña', 76), c('niñeo', 76)]}
        ${'nino'} | ${false}   | ${[c('niño', 1), c('niños', 51), c('niña', 76), c('niñeo', 76)]}
    `(
        'Tests suggestions weighted "$word" ignoreCase: $ignoreCase',
        async ({ word, ignoreCase, expectedWords }) => {
            const trie = await getTrie();
            const wm = weightMap();
            const results = trie.suggestWithCost(word, { numSuggestions: 4, ignoreCase, weightMap: wm });
            expect(results).toEqual(expectedWords);
        },
        { timeout: 5000 },
    );
    test.each`
        wordA     | wordB
        ${'niño'} | ${'niños'}
        ${'nïño'} | ${'niños'}
        ${'nino'} | ${'niña'}
    `('weighted distance "$wordA" $wordB', async ({ wordA, wordB }) => {
        const nWordA = wordA.normalize('NFD');
        const nWordB = wordB.normalize('NFD');

        const wm = createWeightCostCalculator(weightMap());

        const dex = distanceAStarWeightedEx(wordA, wordB, wm);
        expect(formatExResult(dex)).toMatchSnapshot();

        const dexN = distanceAStarWeightedEx(nWordA, nWordB, wm);
        expect(formatExResult(dexN)).toMatchSnapshot();
    });
});

function c(word: string, cost: number) {
    return { word, cost };
}

const sampleWords = ['niño', 'niños', 'niña', 'niñeo', 'dino', 'nido'];

function trieSimple() {
    return parseLinesToDictionaryLegacy(sampleWords);
}

const defaultDictInfo: DictionaryInformation = {
    locale: 'es-ES',
    // cspell:disable-next-line
    alphabet: 'aeroinsctldumpbgfvhzóíjáqéñxyúükwAEROINSCTLDUMPBGFVHZÓÍJÁQÉÑXYÚÜKW',
    suggestionEditCosts: [
        {
            map: '(o$)(os$)(a$)(eo$)',
            replace: 75,
        },
        {
            map: '(o$)(os$)|(a$)(as$)',
            replace: 50,
        },
    ],
};

function weightMap(di: DictionaryInformation = defaultDictInfo): WeightMap {
    return mapDictionaryInformationToWeightMap(di);
}
