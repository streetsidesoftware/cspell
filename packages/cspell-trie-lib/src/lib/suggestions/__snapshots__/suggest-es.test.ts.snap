// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate Spanish Suggestions weighted distance "nino" niña 1`] = `
"<nino> -> <niña> (76)
a: |<^>|<n>|<i>|<n>|<o$>|     |
b: |<^>|<n>|<i>|<ñ>|<a$>|     |
c: |  0|  0|  0|  1|  75| = 76|
p: |  0|  0|  0|  0|   0|  = 0|
"
`;

exports[`Validate Spanish Suggestions weighted distance "nino" niña 2`] = `
"<nino> -> <niña> (76)
a: |<^>|<n>|<i>|<n>|<o$>|     |
b: |<^>|<n>|<i>|<ñ>|<a$>|     |
c: |  0|  0|  0|  1|  75| = 76|
p: |  0|  0|  0|  0|   0|  = 0|
"
`;

exports[`Validate Spanish Suggestions weighted distance "niño" niños 1`] = `
"<niño> -> <niños> (50)
a: |<^>|<n>|<i>|<ñ>|<o$> |     |
b: |<^>|<n>|<i>|<ñ>|<os$>|     |
c: |  0|  0|  0|  0|   50| = 50|
p: |  0|  0|  0|  0|    0|  = 0|
"
`;

exports[`Validate Spanish Suggestions weighted distance "niño" niños 2`] = `
"<niño> -> <niños> (50)
a: |<^>|<n>|<i>|<n>|<̃>|<o$> |     |
b: |<^>|<n>|<i>|<n>|<̃>|<os$>|     |
c: |  0|  0|  0|  0| 0|   50| = 50|
p: |  0|  0|  0|  0| 0|    0|  = 0|
"
`;

exports[`Validate Spanish Suggestions weighted distance "nïño" niños 1`] = `
"<nïño> -> <niños> (150)
a: |<^>|<n>|<ï>|<ñ>|<o$> |      |
b: |<^>|<n>|<i>|<ñ>|<os$>|      |
c: |  0|  0|100|  0|   50| = 150|
p: |  0|  0|  0|  0|    0|   = 0|
"
`;

exports[`Validate Spanish Suggestions weighted distance "nïño" niños 2`] = `
"<nïño> -> <niños> (51)
a: |<^>|<n>|<i>|<̈>|<n>|<̃>|<o$> |     |
b: |<^>|<n>|<i>|<>|<n>|<̃>|<os$>|     |
c: |  0|  0|  0| 1|  0| 0|   50| = 51|
p: |  0|  0|  0| 0|  0| 0|    0|  = 0|
"
`;
